design-documents/checkout-service.md

```diff
@@ -0,0 +1,123 @@
+Possible optimizations: need to investigate if we can eliminate 1 call, as a result we can potentially get 1 call.
+
+Tax
+* 2 calls
+Possible optimizations: need to investigate if we can eliminate 1 call, as a result we will get 1 call.
+
+Integration
+* 1 call that can be removed.
+
+#### Additional Optimizations
+
+* Pass customer/group/addresses as an argument.
+* Remove inventory checks, and check inventory to display whether product is stock when user gets to cart page. There are might be checks on qty increment to make sure that product being added with correct qty increment
+* Totals may be calculated after adding product to cart in a separate operation
+* Totals should not be recalculated on each cart load page, only when cart changes. It will should be recalculated when placing the order to make sure product data didn't change.
+* Create service for totals calculation to reduce number of calls to BFF
+* Group calls to the same service (request all needed data in the beginning); can lead to shared state
+
+After we applying the following optimizations, we will have communication with checkout service as shown below
+* Pass customer as an argument of add to cart operation
+* Aggregate calls to catalog and inventory services Magento\CatalogInventory\Api\StockRegistryInterface::getStockItemBySku, Magento\CatalogInventory\Api\StockRegistryInterface::getStockStatusBySku need to combine into one call
+* Get product info and get final price can be combined into one call
+* Prepare item for cart logic should be eliminated, service should receive cart ready to add
+* Create API that would allow to calculate totals for the quote
+
+![First iteration](checkout-service/add-to-cart-first-iteration.png)
+
+
+If we apply these additional optimizations, communication between services will look like this
+* Have tax service return aggregated information that later can be used to display tax on product page or totals. Question: wouldn't this data be redundant in some cases?
+* Move totals in it's own service: 1) reduce number of calls to monolith, 2) cache of requested data from tax, quote and sales service and as a result reduce amount of communication.
+
+
+![Apply more decomposition](checkout-service/add-to-cart-apply-more-decomposition.png)
+
+Cons
+* More work for initial iteration as we would have to decouple more services with checkout
+
+
+Pros
+* Allows reduce calls to BFF
+* Allows to avoid intermediate state of calculating totals in BFF and reduce amount of backwards incompatible changes
+
+### Place Order Operation
+
+Communication for place order operation looks similar. The only major differences are that we need to save payment, load quote multiple times and save quote. This flow potentially can be optimized.
+
+
+![Apply more decomposition](checkout-service/place-order.png)
+
+
+## Transactions
+Currently Quote module is responsible for creating an order. Need to move this responsibility to order service (BFF after first iteration).
```

Comment:
"BFF" - do you mean "monolith"? Thought it's the same thing after first iteration, the meaning is different. It's strange responsibility for BFF to place an order. Monolith represents services, just deployed on a single instance (with in-memory dependencies), and in this case placing order is an OK responsibility for the monolith.

Reply:
Yes, by BFF I mean monolith. I specified that placing of an order is responsibility of order service. Are you suggesting to reword it?